### PR TITLE
OBPIH-4190 Added Requestor permission

### PR DESCRIPTION
**Pull request by MagdaToczek**

### CODE REVIEW

File: grails-app/taglib/org/pih/warehouse/SelectTagLib.groovy [hunk 1]
```diff
@@ -501,6 +501,30 @@ class SelectTagLib {
...
+    def selectLocationWithSubmitRequestActivity = { attrs, body ->
+
+        // If attrs.from is populated use that by default even if it's empty
+        if (!attrs.containsKey("from")) {
+            ActivityCode activityCode = attrs.activityCode ?: null
+            attrs.from = locationService.getAllLocations().sort { it?.name?.toLowerCase() }
+
+            // use sparingly - this is expensive since it requires multiple database queries
+            if (activityCode) {
+                attrs.from = attrs.from.findAll { it.supports(activityCode) || (!it.supports(activityCode) && it.supports(ActivityCode.SUBMIT_REQUEST)) }
```
Comment (jmiranda) [inline grails-app/taglib/org/pih/warehouse/SelectTagLib.groovy:513]: This doesn't look right. What are we trying to achieve here and why won't this work by generically passing ActivityCode.SUBMIT_REQUEST as the activityCode to selectLocation?

Basically, I'd like to make it so we only need selectLocation. Is there any way to make that happen?

File: grails-app/services/org/pih/warehouse/core/LocationService.groovy [hunk 2]
```diff
@@ -343,6 +352,10 @@ class LocationService {
         return locations.values().toList()
     }
 
+    List getSubmitRequestLocations() {
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/core/LocationService.groovy:355]: Let's call this getRequestingLocations()

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9 http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+
+  <changeSet author="MagdaToczek" id="1412202111400-0">
```
Comment (jmiranda) [inline grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:6]: Can you add a comment to this changeset to explain what's going on, including information about what location type you're targeting (location_type_id = 2)? 

Also we should be using location_type_code = DEPOT instead. But I suspect that makes the insert more difficult, so I think we can leave this as-is.

**Review by jmiranda — CHANGES_REQUESTED** [review]

File: grails-app/taglib/org/pih/warehouse/SelectTagLib.groovy (hunk 1 above)
Comment (MagdaToczek) [reply grails-app/taglib/org/pih/warehouse/SelectTagLib.groovy:513]: Sure, that is leftover from my trials, sorry for that.

File: grails-app/services/org/pih/warehouse/core/LocationService.groovy (hunk 2 above)
Comment (MagdaToczek) [reply grails-app/services/org/pih/warehouse/core/LocationService.groovy]: ok

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml (hunk 3 above)
Comment (MagdaToczek) [reply grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:7]: ok

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9 http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+
+  <!-- selection for location_type_code = 'DEPOT' AND supported_activities_string = 'SUBMIT_REQUEST'-->
```
Comment (awalkowiak) [inline grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:6]: Add a note, that for code depot the id of location type is `2` (if you are leaving the id conditional)

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml (hunk 3 above)
Comment (awalkowiak) [reply grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:7]: @jmiranda why do you think the insert will be more difficult in this case. What if someone changed the id of the location type with code `DEPOT`? I think we can not rely on our inner ids too much.

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        AND supported_activities_string = 'SUBMIT_REQUEST';
+      </sqlCheck>
+    </preConditions>
+    <insert tableName="location_type_supported_activities">
+      <column name="location_type_id" value="2"/>
+      <column name="supported_activities_string" value="SUBMIT_REQUEST"/>
+    </insert>
+  </changeSet>
+
+</databaseChangeLog>
```
Comment (awalkowiak) [inline grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:22]: Add an empty line at the end of file.

**Review by awalkowiak — CHANGES_REQUESTED** [review]
@MagdaToczek Overall LGTM, I am not sure only about the migration

cc @jmiranda

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml (hunk 5 above)
Comment (MagdaToczek) [reply grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:22]: ok

File: grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml (hunk 3 above)
Comment (MagdaToczek) [reply grails-app/migrations/0.8.x/changelog-2021-12-14-1140-add-submit-request-activity.xml:7]: As my excuse I followed on existing migration (changelog-2018-07-09-2222-insert-internal-receiving-location-type.xml)

**Review by pmuchowski — APPROVED** [review]

**Pull request merged**